hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java

```diff
@@ -105,20 +107,27 @@ public static JoinType toHibernateJoinType(int astJoinType) {
 
 	private <T extends AST> List<T> findAllNodes(AST node, Class<T> clazz) {
 		ArrayList<T> found = new ArrayList<>();
-		doFindAllNodes( node, clazz, found );
-		return found;
-	}
 
-	private <T extends AST> void doFindAllNodes(AST node, Class<T> clazz, List<T> found) {
-		if ( clazz.isAssignableFrom( node.getClass() ) ) {
-			found.add( (T) node );
-		}
-		if ( node.getFirstChild() != null ) {
-			doFindAllNodes( node.getFirstChild(), clazz, found );
-		}
-		if ( node.getNextSibling() != null ) {
-			doFindAllNodes( node.getNextSibling(), clazz, found );
+		Deque<AST> nodeDeque = new ArrayDeque<>();
+		nodeDeque.push( node );
+
+		while ( !nodeDeque.isEmpty() ) {
+			AST currentNode = nodeDeque.pop();
+			if ( clazz.isAssignableFrom( currentNode.getClass() ) ) {
+				found.add( (T) currentNode );
+			}
+
+			AST nextSibling = currentNode.getNextSibling();
+			if ( nextSibling != null ) {
+				nodeDeque.push( nextSibling );
+			}
+
+			AST firstChild = currentNode.getFirstChild();
+			if ( firstChild != null ) {
+				nodeDeque.push( firstChild );
+			}
```

Comment:
Seems there is some order difference after the refactoring. Previously nodes are found depth-first, now are found breadth-first. If we swap the above code block #120-123 with code block #125-128, the order should be the same. Not sure whether the node order found matters or not, though.
Overall, pretty good refactoring.